libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,142 @@
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+#include <libevmasm/Exceptions.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+// Walks the call graph using a Depth-First-Search assigning memory offsets to variables.
+// - The leaves of the call graph will get the lowest offsets, increasing towards the root.
+// - ``nextAvailableSlot`` maps a function to the next available slot that can be used by another
+//   function that calls it.
+// - For each function starting from the root of the call graph:
+//   - Visit all children that are not already visited.
+//   - Determe the maximum value ``n`` of the values of ``nextAvailableSlot`` among the children.
+//   - If the function itself contains variables that need memory slots, but is contained in a cycle,
+//     abort the process as failure.
+//   - If not, assign each variable its slot starting starting from ``n`` (incrementing it).
+//   - Assign ``nextAvailableSlot`` of the function to ``n``.
+struct MemoryOffsetAllocator
+{
+	map<YulString, set<YulString>> const& unreachableVariables;
+	map<YulString, set<YulString>> const& callGraph;
+
+	uint64_t run(YulString _function = YulString{})
+	{
+		if (nextAvailableSlot.count(_function))
+			return nextAvailableSlot[_function];
+
+		// Assign to zero early to guard against recursive calls.
+		nextAvailableSlot[_function] = 0;
+
+		uint64_t nextSlot = 0;
+		if (callGraph.count(_function))
+			for (YulString child: callGraph.at(_function))
+				nextSlot = std::max(run(child), nextSlot);
+
+		if (unreachableVariables.count(_function))
+		{
+			yulAssert(!slotAllocations.count(_function), "");
+			auto& assignedSlots = slotAllocations[_function];
+			for (YulString variable: unreachableVariables.at(_function))
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+					assignedSlots[variable] = nextSlot++;
+		}
+
+		return (nextAvailableSlot[_function] = nextSlot);
+	}
+
+	map<YulString, map<YulString, uint64_t>> slotAllocations{};
+	map<YulString, uint64_t> nextAvailableSlot{};
+};
+u256 getLiteralArgumentValue(FunctionCall const& _call)
+{
+	yulAssert(_call.arguments.size() == 1, "");
+	Literal const* literal = std::get_if<Literal>(&_call.arguments.front());
+	yulAssert(literal, "");
+	return valueOfLiteral(*literal);
+}
+}
+
+void StackLimitEvader::run(
+	OptimiserStepContext& _context,
+	Object& _object,
+	map<YulString, set<YulString>> const& _unreachableVariables
+)
+{
+	yulAssert(_object.code, "");
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	vector<FunctionCall*> getFreeMemoryStartCalls = FunctionCallFinder::run(
+		*_object.code,
+		"memoryguard"_yulstring
+	);
+	if (getFreeMemoryStartCalls.empty())
+		return;
```

Comment:
Leave a comment that "Do not optimise if no memoryguard found".

Reply:
I'll also rename the variable while I'm at it :-).